### PR TITLE
KAFKA-9505: Only loop over topics-to-validate in retries

### PR DESCRIPTION
**Pull request by guozhangwang**
Found this bug from the repeated flaky runs of system tests, it seems to be long lurking but also would only happen if there are frequent rebalances / topic creation within a short time, which is exactly the case in some of our smoke system tests.

Also added a unit test.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/InternalTopicManagerTest.java [hunk 1]
```diff
@@ -132,7 +141,49 @@ public void shouldCreateRequiredTopics() throws Exception {
         assertEquals(new ConfigEntry(TopicConfig.CLEANUP_POLICY_CONFIG, TopicConfig.CLEANUP_POLICY_DELETE), mockAdminClient.describeConfigs(Collections.singleton(resource)).values().get(resource).get().get(TopicConfig.CLEANUP_POLICY_CONFIG));
         assertEquals(new ConfigEntry(TopicConfig.CLEANUP_POLICY_CONFIG, TopicConfig.CLEANUP_POLICY_COMPACT), mockAdminClient.describeConfigs(Collections.singleton(resource2)).values().get(resource2).get().get(TopicConfig.CLEANUP_POLICY_CONFIG));
         assertEquals(new ConfigEntry(TopicConfig.CLEANUP_POLICY_CONFIG, TopicConfig.CLEANUP_POLICY_COMPACT + "," + TopicConfig.CLEANUP_POLICY_DELETE), mockAdminClient.describeConfigs(Collections.singleton(resource3)).values().get(resource3).get().get(TopicConfig.CLEANUP_POLICY_CONFIG));
+    }
 
+    @Test
+    public void shouldRetryValidatingTopics() {
```
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/InternalTopicManagerTest.java]: This added test will fail without the fix.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java [hunk 2]
```diff
@@ -217,15 +216,17 @@ public void makeReady(final Map<String, InternalTopicConfig> topics) {
...
+        }
 
         final Map<String, Integer> existedTopicPartition = getNumPartitions(topicsToValidate);
 
         final Set<String> topicsToCreate = new HashSet<>();
-        for (final Map.Entry<String, InternalTopicConfig> entry : topicsMap.entrySet()) {
-            final String topicName = entry.getKey();
-            final Optional<Integer> numberOfPartitions = entry.getValue().numberOfPartitions();
+        for (final String topicName : topicsToValidate) {
+            final Optional<Integer> numberOfPartitions = topicsMap.get(topicName).numberOfPartitions();
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java:229]: This is the actual fix: we should not loop over the global topic-map but only the remaining `topics-to-validate`. cc @ableegoldman

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java [hunk 3]
```diff
@@ -1037,7 +1037,7 @@ private long sendEligibleCalls(long now) {
                 }
                 ClientRequest clientRequest = client.newClientRequest(node.idString(), requestBuilder, now,
                         true, requestTimeoutMs, null);
-                log.trace("Sending {} to {}. correlationId={}", requestBuilder, node, clientRequest.correlationId());
+                log.debug("Sending {} to {}. correlationId={}", requestBuilder, node, clientRequest.correlationId());
```
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:1040]: I found this debug line very crucial; and in other clients (producer, consumer) the same line indicating when we actual put the request to the send queue was set at debug level as well, so I've changed it to also debug.

**Review by guozhangwang — COMMENTED** [review]
cc @ableegoldman @vvcephei for reviews.

File: clients/src/main/java/org/apache/kafka/clients/admin/CreateTopicsResult.java [hunk 4]
```diff
@@ -35,7 +35,7 @@
 
     private final Map<String, KafkaFuture<TopicMetadataAndConfig>> futures;
 
-    CreateTopicsResult(Map<String, KafkaFuture<TopicMetadataAndConfig>> futures) {
+    public CreateTopicsResult(Map<String, KafkaFuture<TopicMetadataAndConfig>> futures) {
```
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/admin/CreateTopicsResult.java]: I'd have to make these constructors public in order to mock the response in the unit test

**Comment by guozhangwang** [comment]
retest this please

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java [hunk 5]
```diff
@@ -217,15 +216,17 @@ public void makeReady(final Map<String, InternalTopicConfig> topics) {
     /**
      * Check the existing topics to have correct number of partitions; and return the remaining topics that needs to be created
      */
-    private Set<String> validateTopics(final Set<String> topicsToValidate,
-                                       final Map<String, InternalTopicConfig> topicsMap) {
+    private Set<String> validateTopics(final Set<String> topicsToValidate, final Map<String, InternalTopicConfig> topicsMap) {
+        if (!topicsMap.keySet().containsAll(topicsToValidate)) {
```
Comment (abbccdda) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java:220]: Do we still need to pass in the topicsMap, as we are no longer looping it?

Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java:220]: Yes we need the InternalTopicConfig object from it :)

Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java:220]: More specifically we need the `numberOfPartitions` to compare.

**Comment by guozhangwang** [comment]
retest this please

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java [hunk 6]
```diff
@@ -217,15 +216,17 @@ public void makeReady(final Map<String, InternalTopicConfig> topics) {
...
+            throw new IllegalStateException("The topics map " + topicsMap + " does not contain all the topics " +
```
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java]: req: Can we log an error with the two topics sets (`topicsMap.keySet` and `topicsToValidate`), and then just say what happened in the exception message? Logging the entire map in the exception message will be pretty ugly

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/InternalTopicManagerTest.java (hunk 1 above)
Comment (ableegoldman) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/InternalTopicManagerTest.java]: nit: Can we rename to `topicValidationShouldCompleteOnRetry` (or something)?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java (hunk 2 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java:229]: Yikes, nice find! I'm pretty sure this is behind that C3 issue. I'm sure there's a good reason for them to race against Streams in the internal topic creation...

**Review by ableegoldman — CHANGES_REQUESTED** [review]
LGTM, modulo the one comment about logging

**Comment by guozhangwang** [comment]
retest this please

**Comment by ableegoldman** [comment]
@guozhangwang build failed on spotBugsMain

**Comment by guozhangwang** [comment]
I noticed that too, it is from `examples` repo and seems caused by a bad commit before, will try to push a hotfix commit.

**Comment by guozhangwang** [comment]
retest this please

**Review by ableegoldman — APPROVED** [review]
LGTM assuming the tests pass

**Comment by guozhangwang** [comment]
retest this please

**Comment by guozhangwang** [comment]
retest this please

**Comment by guozhangwang** [comment]
retest this please

File: clients/src/main/java/org/apache/kafka/clients/admin/CreateTopicsResult.java (hunk 4 above)
Comment (mjsax) [reply clients/src/main/java/org/apache/kafka/clients/admin/CreateTopicsResult.java]: What about making it protected instead and mock a sub-class? `CreateTopicsResults` is part of the public API and we might not want to make them public (would this be a public API change?)

**Review by mjsax — COMMENTED** [review]
Nice find! Overall LGTM. I am not sure about making the constructors public though.

File: clients/src/main/java/org/apache/kafka/clients/admin/CreateTopicsResult.java (hunk 4 above)
Comment (guozhangwang) [reply clients/src/main/java/org/apache/kafka/clients/admin/CreateTopicsResult.java]: Yeah that's a good idea! Let me try that out.

**Comment by guozhangwang** [comment]
@mjsax I changed the constructor back as you suggested. If you feel good about the current PR I'll go ahead and merge it.

File: clients/src/main/java/org/apache/kafka/clients/admin/CreateTopicsResult.java [hunk 7]
```diff
@@ -94,7 +94,7 @@
         return futures.get(topic).thenApply(TopicMetadataAndConfig::replicationFactor);
     }
 
-    static class TopicMetadataAndConfig {
+    public static class TopicMetadataAndConfig {
```
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/admin/CreateTopicsResult.java:97]: This still needs to be public: I looked at other similar objects in XXXResult in `admin`.

**Review by mjsax — APPROVED** [review]
Thanks @guozhangwang. LGTM. Feel free to merge after the build passed.

**Comment by guozhangwang** [comment]
Test failures are relevant to https://github.com/apache/kafka/pull/8081, @abbccdda has a PR already for reducing its scope. I will go ahead and merge it and cherry-pick to 2.5

**Pull request merged**